core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -278,35 +282,38 @@ private[spark] class Executor(
   }
 
   def stop(): Unit = {
-    env.metricsSystem.report()
-    try {
-      metricsPoller.stop()
-    } catch {
-      case NonFatal(e) =>
-        logWarning("Unable to stop executor metrics poller", e)
-    }
-    try {
-      heartbeater.stop()
-    } catch {
-      case NonFatal(e) =>
-        logWarning("Unable to stop heartbeater", e)
-    }
-    threadPool.shutdown()
-
-    // Notify plugins that executor is shutting down so they can terminate cleanly
-    Utils.withContextClassLoader(replClassLoader) {
-      executorPlugins.foreach { plugin =>
-        try {
-          plugin.shutdown()
-        } catch {
-          case e: Exception =>
-            logWarning("Plugin " + plugin.getClass().getCanonicalName() + " shutdown failed", e)
+    if (!executorShutdown) {
```

Comment:
Although this is unlikely to be a problem, this is not thread-safe. You should use an `AtomicBoolean`. Or make `stop()` synchronized.

Reply:
executorShutdown is volatile; the race is still that it could be set between this line and the next? yeah agree. While we're here, worth fixing with AtomicBoolean and compareAndSet